pkg/controller/seed-controller-manager/initialmachinedeployment/controller.go

```diff
@@ -0,0 +1,243 @@
+	reconciler := &Reconciler{
+		Client: mgr.GetClient(),
+
+		ctx:                           ctx,
+		workerName:                    workerName,
+		recorder:                      mgr.GetEventRecorderFor(ControllerName),
+		seedGetter:                    seedGetter,
+		userClusterConnectionProvider: userClusterConnectionProvider,
+		log:                           log,
+		versions:                      versions,
+	}
+
+	c, err := controller.New(ControllerName, mgr, controller.Options{
+		Reconciler:              reconciler,
+		MaxConcurrentReconciles: numWorkers,
+	})
+	if err != nil {
+		return fmt.Errorf("failed to create controller: %v", err)
+	}
+
+	if err := c.Watch(&source.Kind{Type: &kubermaticv1.Cluster{}}, &handler.EnqueueRequestForObject{}); err != nil {
+		return fmt.Errorf("failed to create watch: %v", err)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	cluster := &kubermaticv1.Cluster{}
+	if err := r.Get(r.ctx, request.NamespacedName, cluster); err != nil {
+		if kerrors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+
+	// Add a wrapping here so we can emit an event on error
+	result, err := kubermaticv1helper.ClusterReconcileWrapper(
+		r.ctx,
+		r.Client,
+		r.workerName,
+		cluster,
+		r.versions,
+		kubermaticv1.ClusterConditionMachineDeploymentControllerReconcilingSuccess,
+		func() (*reconcile.Result, error) {
+			return r.reconcile(cluster)
+		},
+	)
+	if err != nil {
+		r.log.Errorw("Failed to reconcile cluster", "cluster", cluster.Name, zap.Error(err))
+		r.recorder.Event(cluster, corev1.EventTypeWarning, "ReconcilingError", err.Error())
+	}
+	if result == nil {
+		result = &reconcile.Result{}
+	}
+	return *result, err
+}
+
+func (r *Reconciler) reconcile(cluster *kubermaticv1.Cluster) (*reconcile.Result, error) {
+	// If cluster is not healthy yet there is nothing to do.
+	// If it gets healthy we'll get notified by the event. No need to requeue.
+	if !cluster.Status.ExtendedHealth.AllHealthy() {
+		return nil, nil
+	}
+
+	request := cluster.Annotations[v1.InitialMachineDeploymentRequestAnnotation]
+
+	// if there is no (more) annotation, we're done
+	if request == "" {
+		return nil, nil
+	}
+
+	var nodeDeployment *v1.NodeDeployment
+	if err := json.Unmarshal([]byte(request), &nodeDeployment); err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
+		return nil, fmt.Errorf("cannot unmarshal initial MachineDeployment request: %v", err)
+	}
+
+	nodeDeployment, err := machineresource.Validate(nodeDeployment, cluster.Spec.Version.Semver())
+	if err != nil {
+		// TODO: What now? Remove the invalid annotation? This will end up in a reconcile loop.
```

Comment:
I think that removing the annotation in both these cases makes sense since there is really no obvious way to fix these issues here. I would just leave some information in the event that creation has failed and remove the annotation.

Reply:
All the errors go to event recorder already, so removing the annotation in that case is good way to go for me.